airgun/widgets.py

```diff
@@ -489,39 +513,58 @@ class CustomParameter(Widget):
         //textarea[@placeholder='Value']
 
     """
-    add_new_value = Text(".//a[contains(text(),'+ Add Parameter')]")
-    new_parameter_name = TextInput(
-        locator=".//input[@placeholder='Name' and not(@value)]")
-    new_parameter_value = TextInput(
-        locator=".//textarea[@placeholder='Value' and not(text())]")
-    NAMES = (
-        ".//tr[contains(@id, 'os_parameter')]/td/input[@placeholder='Name']")
-    VALUE = (
-        ".//table[contains(@id, 'parameters')]//tr"
-        "/td[input[contains(@id, 'name')][contains(@value, '{}')]]"
-        "/following-sibling::td//textarea"
-    )
+    add_new_value = Text("..//a[contains(text(),'+ Add Parameter')]")
+
+    def __init__(self, *args, **kwargs):
+        """Supports initialization via ``locator=`` or ``id=``"""
+        locator = kwargs.get('locator')
+        id = kwargs.pop('id')
+        if locator and id or not locator and not id:
+            raise ValueError('Please specify either locator or id')
+        kwargs['locator'] = locator or ".//table[@id='{}']".format(id)
+
+        column_widgets = {
+            'Name': TextInput(locator=".//input[@placeholder='Name']"),
+            'Value': TextInput(locator=".//textarea[@placeholder='Value']"),
+            'Actions': Text(locator=".//a[@title='Remove Parameter']")
+        }
+        super(CustomParameter, self).__init__(
+            *args, column_widgets=column_widgets, **kwargs)
 
     def read(self):
         """Return a list of dictionaries. Each dictionary consists of name and
         value parameters
         """
         parameters = []
-        for item in self.browser.elements(self.NAMES):
-            name = self.browser.get_attribute('value', item)
-            value = self.browser.text(self.VALUE.format(name))
+        for row in self.rows():
+            name = row['Name'].widget.read()
+            value = row['Value'].widget.read()
             parameters.append({'name': name, 'value': value})
         return parameters
 
-    def fill(self, values):
-        """Create new parameter entity
+    def add(self, value):
+        """Add single name/value parameter entry to the table.
 
-        :param values: dictionary of name and value that should be assigned to
-            new entity
+        :param value: dict with format {'name': str, 'value': str}
         """
         self.add_new_value.click()
-        self.new_parameter_name.fill(values['name'])
-        self.new_parameter_value.fill(values['value'])
+        new_row = self.__getitem__(-1)
+        new_row['Name'].widget.fill(value['name'])
+        new_row['Value'].widget.fill(value['value'])
+
+    def fill(self, values):
+        """Fill parameter entries. Existing values will be overwritten.
```

Comment:
This change may effect existing tests, we need to make sure they're still working as expected before merging.